### PR TITLE
fix(provision/gce): add zone fallback on GCE resource pool exhaustion

### PR DESCRIPTION
**Pull request by juliayakovlev**
When a GCE zone is exhausted (ZONE_RESOURCE_POOL_EXHAUSTED), ZoneResourcesExhaustedError was raised but never caught, crashing the test. Add zone-level fallback in GCECluster._create_instances() that tries alternative zones in the same region before giving up. Also expand us-central1 from single zone 'a' to 'abcf'.

Fixes: https://scylladb.atlassian.net/browse/SCT-350

### Testing

- [x] [artifacts-ubuntu2204-test](https://argus.scylladb.com/tests/scylla-cluster-tests/12caffa6-340b-4b22-b461-4ea3f5a6d7c0)

I ran the test with simulation of the failure. The instance was provided with fallback into another AZ
```
19:33:24  < t:2026-05-18 16:33:23,847 f:cluster_gce.py  l:555  c:sdcm.cluster         p:INFO  > GCE Cluster artifacts-ubuntu2204-jenkins-db-cluster-12caffa6 | Image: ubuntu-2204-lts | Root Disk: pd-ssd 50 GB | Local SSD: 1 | : Found no provisioned instances. Provision them.
19:33:25  < t:2026-05-18 16:33:24,120 f:cluster_gce.py  l:398  c:sdcm.cluster         p:WARNING > GCE Cluster artifacts-ubuntu2204-jenkins-db-cluster-12caffa6 | Image: ubuntu-2204-lts | Root Disk: pd-ssd 50 GB | Local SSD: 1 | : Zone us-east1-c exhausted, trying alternative zones in region us-east1
19:33:25  < t:2026-05-18 16:33:24,121 f:cluster_gce.py  l:403  c:sdcm.cluster         p:INFO  > GCE Cluster artifacts-ubuntu2204-jenkins-db-cluster-12caffa6 | Image: ubuntu-2204-lts | Root Disk: pd-ssd 50 GB | Local SSD: 1 | : Attempting zone us-east1-d as fallback
19:33:25  < t:2026-05-18 16:33:24,682 f:instance_provider.py l:120  c:sdcm.provision.gce.instance_provider p:INFO  > Creating 'artifacts-ubuntu2204-jenkins-db-node-12caffa6-0-1' VM in zone us-east1-d...
19:33:25  < t:2026-05-18 16:33:24,682 f:instance_provider.py l:121  c:sdcm.provision.gce.instance_provider p:INFO  > Instance params: InstanceDefinition(name='artifacts-ubuntu2204-jenkins-db-node-12caffa6-0-1', image_id='https://www.googleapis.com/compute/v1/projects/ubuntu-os-cloud/global/images/family/ubuntu-2204-lts', type='n2-standard-2', user_name='scylla-test', tags={'RunByUser': 'julia.yakovlev', 'TestName': 'artifacts_test.ArtifactsTest.test_scylla_service', 'TestId': '12caffa6-340b-4b22-b461-4ea3f5a6d7c0', 'version': 'scylla-staging', 'CreatedBy': 'SCT', 'JenkinsJobTag': 'jenkins-scylla-staging-yulia-artifacts-ubuntu2204-test-25', 'JenkinsJob': 'jenkins-scylla-staging-yulia-artifacts-ubuntu2204-test', 'billing_project': 'no_billing_project', 'NodeType': 'scylla-db', 'keep_action': 'terminate', 'NodeIndex': '1'}, arch=<VmArch.X86: 'x86_64'>, root_disk_size=50, root_disk_type='pd-ssd', data_disks=[DataDisk(type='local-ssd', size=375, iops=0, count=1)], use_public_ip=False, rack_index=None)
19:33:25  < t:2026-05-18 16:33:24,819 f:instance_provider.py l:287  c:sdcm.provision.gce.instance_provider p:INFO  > Creating instance artifacts-ubuntu2204-jenkins-db-node-12caffa6-0-1 in zone us-east1-d...
19:33:36  < t:2026-05-18 16:33:36,011 f:instance_provider.py l:166  c:sdcm.provision.gce.instance_provider p:INFO  > Instance artifacts-ubuntu2204-jenkins-db-node-12caffa6-0-1 created successfully
```

- [x] [arm artifact test](https://argus.scylladb.com/tests/scylla-cluster-tests/dcd4302c-2203-4a69-be59-70ab98230d82)
- [x] [arm artifact test with fallback simulation](https://argus.scylladb.com/tests/scylla-cluster-tests/98c269fb-5c92-466b-b586-68cb61183e0f)
```
14:43:34  < t:2026-05-25 11:43:33,542 f:cluster_gce.py  l:408  c:sdcm.cluster         p:WARNING > GCE Cluster artifacts-ubuntu2204-jenkins-db-cluster-98c269fb | Image: ubuntu-2204-lts | Root Disk: pd-ssd 50 GB | Local SSD: 1 | : Zone us-central1-c exhausted, trying alternative zones in region us-central1
14:43:34  < t:2026-05-25 11:43:34,069 f:cluster_gce.py  l:433  c:sdcm.cluster         p:INFO  > GCE Cluster artifacts-ubuntu2204-jenkins-db-cluster-98c269fb | Image: ubuntu-2204-lts | Root Disk: pd-ssd 50 GB | Local SSD: 1 | : Attempting zone us-central1-a as fallback
```

- [x] [artifacts-ubuntu2604-arm-test with fallback simulation](https://argus.scylladb.com/tests/scylla-cluster-tests/d945e48b-7619-4458-85e4-efe1fd2553b3)
```
15:26:16  < t:2026-05-25 12:26:16,413 f:cluster_gce.py  l:408  c:sdcm.cluster         p:WARNING > GCE Cluster artifacts-ubuntu2604-jenkins-db-cluster-d945e48b | Image: ubuntu-2604-lts-arm64 | Root Disk: pd-ssd 50 GB | pd-standard: 50 | : Zone us-central1-f exhausted, trying alternative zones in region us-central1
15:26:18  < t:2026-05-25 12:26:17,180 f:cluster_gce.py  l:433  c:sdcm.cluster         p:INFO  > GCE Cluster artifacts-ubuntu2604-jenkins-db-cluster-d945e48b | Image: ubuntu-2604-lts-arm64 | Root Disk: pd-ssd 50 GB | pd-standard: 50 | : Attempting zone us-central1-a as fallback
```

- [x] [artifacts-gce-image-test with fallback simulation](https://argus.scylladb.com/tests/scylla-cluster-tests/00665c9b-735e-4f91-9c25-c4d282a397ad)
```
16:21:22  < t:2026-05-25 13:21:21,992 f:cluster_gce.py  l:408  c:sdcm.cluster         p:WARNING > GCE Cluster artifacts-gce-image-jenkins-db-cluster-00665c9b | Image: 4903389776318144561 | Root Disk: pd-ssd 50 GB | Local SSD: 8 | : Zone us-central1-f exhausted, trying alternative zones in region us-central1
16:21:22  < t:2026-05-25 13:21:22,643 f:cluster_gce.py  l:433  c:sdcm.cluster         p:INFO  > GCE Cluster artifacts-gce-image-jenkins-db-cluster-00665c9b | Image: 4903389776318144561 | Root Disk: pd-ssd 50 GB | Local SSD: 8 | : Attempting zone us-central1-a as fallback
```

- [x] [artifacts-gce-image](https://argus.scylladb.com/tests/scylla-cluster-tests/c6424290-4a3c-4e27-81c5-5d13e4eb4cb6)
- [x] [rolling-upgrade-ubuntu2404-test](https://argus.scylladb.com/tests/scylla-cluster-tests/2696c394-85fc-4cf4-bce4-777b14452749)

### PR pre-checks (self review)

- [x] I added the relevant `backport` labels
- [x] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

### CODE REVIEW

**Comment by fruch** [comment]
## Code Review

### Issues Found: 2

#### 1. 🐛 `test_returns_empty_when_only_one_zone` — test name and logic contradict each other

**File**: [`unit_tests/unit/test_gce_zone_fallback.py`](https://github.com/scylladb/scylla-cluster-tests/blob/86eacd8e7105281224789ef54a1eb8a68f821965/unit_tests/unit/test_gce_zone_fallback.py#L36-L40)

The test name says it verifies empty results when only one zone exists, but:
- It uses `us-east1` which has **2 zones** (`"cd"`), not 1
- It asserts `len(alternatives) >= 1` — the **opposite** of empty

The actual edge case (single-zone region returning `[]`) is completely untested. The test passes vacuously. Consider something like:

```python
def test_returns_empty_when_only_one_zone(self, monkeypatch):
    """If a region has only one zone and it's exhausted, return empty."""
    monkeypatch.setitem(SUPPORTED_REGIONS, "test-region", "a")
    alternatives = get_alternative_zones("test-region", "a")
    assert alternatives == []
```

#### 2. ⚠️ Missing `provision-gce` label

Per [CLAUDE.md](https://github.com/scylladb/scylla-cluster-tests/blob/86eacd8e7105281224789ef54a1eb8a68f821965/CLAUDE.md) — PRs modifying GCE backend files (`sdcm/cluster_gce.py`, `sdcm/utils/gce_utils.py`) should carry the `provision-gce` label to trigger GCE-specific CI jobs. Currently only has `P2` and `ai-assisted`.

---

### Testing Suggestions

To validate the zone fallback logic on real infrastructure, consider using `utils/staging_trigger` to run a few GCE-based artifact tests. Relevant configs:

- `test-cases/artifacts/gce-image.yaml`
- `test-cases/artifacts/gce-image-persistent-disk.yaml`

These use GCE backend with spot instances and `instance_provision_fallback_on_demand: true` — exactly the code path this PR modifies.

Also worth cross-checking: all GCE-based artifact configs (`gce-image.yaml`, `gce-image-persistent-disk.yaml`, and the OS-specific ones like `ubuntu2404.yaml`, `debian12.yaml`, `rocky9.yaml`, etc.) don't specify an explicit region, so they'll use the default region. It would be good to verify that the default region is covered by the expanded `SUPPORTED_REGIONS` map (specifically `us-central1: "abcf"`) and that the zone fallback logic will kick in correctly for these test configurations.

---

Checked for bugs and CLAUDE.md compliance.

File: sdcm/utils/gce_utils.py [hunk 1]
```diff
@@ -108,6 +108,26 @@ def random_zone(region: str) -> str:
...
+    Returns:
+        List of alternative zone letters, shuffled randomly.
+    """
+    availability_zones = SUPPORTED_REGIONS.get(region, "")
+    if not availability_zones:
+        return []
+    # Extract zone letter from full zone name if needed
+    exhausted_letter = exhausted_zone[-1] if len(exhausted_zone) > 1 else exhausted_zone
+    alternatives = [z for z in availability_zones if z != exhausted_letter]
+    random.shuffle(alternatives)
```
Comment (fruch) [inline sdcm/utils/gce_utils.py]: not sure if we want to shuffle them...

File: sdcm/cluster_gce.py [hunk 2]
```diff
@@ -382,12 +383,60 @@ def _create_instances(self, count, dc_idx=0, instance_type=None):
...
+            original_provisioner = self.provisioners[dc_idx]
+            for definition in definitions:
+                if definition.name in original_provisioner._cache:
+                    self.log.info(
+                        "Cleaning up partially-created instance %s in exhausted zone %s",
+                        definition.name,
+                        exhausted_zone,
+                    )
+                    try:
+                        original_provisioner.terminate_instance(definition.name, wait=True)
```
Comment (fruch) [inline sdcm/cluster_gce.py]: this is only fits when we have one instance (i.e. artifacts tests)
we need to introduce `fallback_to_next_availability_zone` (same as `aws_fallback_to_next_availability_zone`) 

for this PR it should cover only 1 node cases, otherwise we need to clear all of the resources when falling back to next AZ

**Comment by fruch** [comment]
## 🔍 Code Review — PR #14616

**fix(provision/gce): add zone fallback on GCE resource pool exhaustion**

### ✅ Checklist Results

| Check | Status | Notes |
|-------|--------|-------|
| Override Safety | ✅ Pass | `_create_instances` signature unchanged — no breakage risk |
| Import Conventions | ✅ Pass | `ZoneResourcesExhaustedError` and `get_alternative_zones` imported at top of file, correct grouping |
| Error Handling | ⚠️ See below | Cleanup errors swallowed; orphan risk |
| Test Coverage | ⚠️ Partial | `get_alternative_zones` well-tested; fallback flow in `_create_instances` not tested |
| Configuration | ✅ N/A | No new config options |
| Backend Impact | ✅ GCE-only | Other backends have their own exhaustion handling |
| Commit Message | ✅ Pass | `fix(provision/gce): ...` — valid Conventional Commits |
| HTTP Resilience | ✅ N/A | No curl/requests changes |

---

### 🔴 Critical Issues (1)

**1. Accessing private `_cache` across class boundaries** — `cluster_gce.py` line ~400:
```python
if definition.name in original_provisioner._cache:
```
`_cache` is an internal detail of `VirtualMachineProvider`. If its structure changes, this breaks silently. 

**Suggestion**: Expose a public method on `GceProvisioner` like `has_cached_instance(name: str) -> bool` or `cached_instance_names() -> set[str]`.

---

### 🟡 Important Issues (3)

**1. Orphaned VMs from partially-sent insert requests**

In `instance_provider.py`, `ZoneResourcesExhaustedError` is raised *during the insert loop* (line 128) when one insert fails. Instances whose insert was already sent but not yet awaited won't be in `_cache` (added at line 167/209 after wait). The cleanup loop in `cluster_gce.py` checks `_cache`, so these will be missed → orphaned VMs incurring cost.

**Suggestion**: Either (a) await all pending operations before raising, collecting successes into `_cache`, or (b) track pending operation names separately so cleanup can cancel/delete them.

**2. Cleanup errors are swallowed silently** — line ~407:
```python
except (google.api_core.exceptions.GoogleAPIError, RuntimeError) as cleanup_err:
    self.log.warning("Failed to clean up instance %s: %s", definition.name, cleanup_err)
```
Failed cleanup = leaked resource = ongoing cost. Consider collecting failed cleanup names and logging them at ERROR level, or adding them to a "leaked resources" report.

**3. Log message inconsistency for zone names** — line ~398:
```python
self.log.warning("Zone %s exhausted, trying alternative zones in region %s", exhausted_zone, region)
```
`exhausted_zone` is `self.provisioners[dc_idx].availability_zone` which is a zone letter (e.g., `"c"`), not a full zone name. But the fallback log uses `f"{region}-{alt_zone}"`. The first log should also format the full zone name for consistency and debuggability.

---

### 💡 Suggestions (2)

1. **Add a unit test for the fallback flow in `_create_instances`** — mock provisioners to verify: cleanup attempted, alternative zones tried in order, provisioner/zone state updated on success, all-exhausted raises correctly. The current test only covers the utility function.

2. **`SUPPORTED_REGIONS` type clarity** — values are strings treated as `Sequence[char]`. A comment or type annotation (`dict[str, str]  # each char is a zone letter`) would prevent confusion if multi-char identifiers are ever needed.

---

### ✅ Strengths

- Clean architectural separation: detection → utility → fallback logic
- Reuses existing `provision_instances_with_fallback` for each retry attempt
- Unit tests are well-structured (pytest style, parametrize, monkeypatch for shuffle)
- PR includes real test evidence with logs showing successful fallback
- `us-central1` zone expansion (`"a"` → `"abcf"`) is a good operational improvement

---

### Verdict

**Approve with minor changes requested.** The critical `_cache` access should be fixed before merge. The orphan-VM risk (important #1) should at least be documented as a known limitation or addressed.

**Comment by juliayakovlev** [comment]
@fruch is this PR relevant? I see your https://github.com/scylladb/scylla-cluster-tests/pull/14619

**Comment by fruch** [comment]
> @fruch is this PR relevant? I see your https://github.com/scylladb/scylla-cluster-tests/pull/14619

Yes it is, my PR isn't for handling the fallback, it suppose to handle the upfront checks only

File: sdcm/utils/gce_utils.py (hunk 1 above)
Comment (juliayakovlev) [reply sdcm/utils/gce_utils.py]: @fruch What do you recommend? Should I always choose the first option?

Comment (fruch) [reply sdcm/utils/gce_utils.py]: we are gonna try them one by one, until one of them would work.
not sure shuffle won't change anything

we should always start from what configured, i.e. if a is configured, we start from it, and to the next, if we can't provision. the order doesn't have lots of meaning, i.e. not point of shuffling.

File: sdcm/cluster_gce.py (hunk 2 above)
Comment (juliayakovlev) [reply sdcm/cluster_gce.py]: done

File: sdcm/cluster_gce.py [hunk 3]
```diff
@@ -382,12 +383,46 @@ def _create_instances(self, count, dc_idx=0, instance_type=None):
...
+                    self._gce_zone_names[dc_idx] = new_provisioner.availability_zone
+                    self.log.info("Successfully provisioned instances in fallback zone %s-%s", region, alt_zone)
+                    return result
+                except ZoneResourcesExhaustedError:
+                    self.log.warning("Fallback zone %s-%s also exhausted, trying next", region, alt_zone)
+                    continue
+            raise ZoneResourcesExhaustedError(
+                f"All zones in region {region} are exhausted: tried {exhausted_zone} and "
+                f"{[f'{region}-{z}' for z in alternative_zones]}"
+            )
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/cluster_gce.py]: done

**Comment by juliayakovlev** [comment]
GCE provisioning failed with SpotTerminationEvent

File: sdcm/cluster_gce.py [hunk 4]
```diff
@@ -382,12 +384,80 @@ def _create_instances(self, count, dc_idx=0, instance_type=None):
...
+                definitions=definitions,
+                pricing_model=pricing_model,
+                fallback_on_demand=self.params.get("instance_provision_fallback_on_demand"),
+            )
+        except (ZoneResourcesExhaustedError, ProvisionError) as exc:
+            if isinstance(exc, ProvisionError) and not _is_zone_exhausted(exc):
+                raise
+            if not (
+                self.params.get("fallback_to_next_availability_zone")
+                or self.params.get("aws_fallback_to_next_availability_zone")
```
Comment (fruch) [inline sdcm/cluster_gce.py]: let's align for deprecation of `aws_fallback_to_next_availability_zone`

Comment (fruch) [reply sdcm/cluster_gce.py]: in a separate commit, let's enable `fallback_to_next_availability_zone` for all gce base artifacts

Comment (juliayakovlev) [reply sdcm/cluster_gce.py]: @fruch it's True by default (in the test_defaults.yaml)

Comment (juliayakovlev) [reply sdcm/cluster_gce.py]: done

File: defaults/test_default.yaml [hunk 5]
```diff
@@ -213,7 +213,7 @@ bare_loaders: false
 
 aws_fallback_to_next_availability_zone: false
 
-fallback_to_next_availability_zone: true
+fallback_to_next_availability_zone: false
```
Comment (fruch) [inline defaults/test_default.yaml:216]: @juliayakovlev please put 

```
fallback_to_next_availability_zone: true
```
in defaults/aws_config.yaml

so it wouldn't revert the behavior change @dimakr did

File: sdcm/cluster_gce.py (hunk 4 above)
Comment (fruch) [reply sdcm/cluster_gce.py]: drop the ref to `aws_fallback_to_next_availability_zone` from here

File: sdcm/cluster_gce.py [hunk 6]
```diff
@@ -382,12 +384,80 @@ def _create_instances(self, count, dc_idx=0, instance_type=None):
...
+            # Clean up any instances that were partially created in the exhausted zone
+            # before retrying in a different zone to avoid duplicates.
+            # Note: ZoneResourcesExhaustedError may escape before the provisioner-level
+            # cache is populated (it's set after _vm_provider.get_or_create returns),
+            # so we also check the underlying VM provider cache.
+            original_provisioner = self.provisioners[dc_idx]
+            for definition in definitions:
+                name = definition.name
+                in_provisioner_cache = name in original_provisioner._cache
+                in_vm_cache = name in original_provisioner._vm_provider._cache
```
[the remark this answers is not in the document]
Comment (fruch) [reply sdcm/cluster_gce.py]: @juliayakovlev 
it's has a point, that you are building ontop of a cache implemented internal at some other part.
it should have a clear public api for it.

or we might have a local cache ? 

also we  currently support only artifact tests, why do we need lookups in any cache ?

[the remark this answers is not in the document]
Comment (fruch) [reply sdcm/cluster_gce.py]: maybe we should waiver it to when we'll need a provision full cluster support with fallback of AZ ?

File: defaults/test_default.yaml (hunk 5 above)
Comment (juliayakovlev) [reply defaults/test_default.yaml:216]: Done.
@dimakr please, confirm that this change won't break your change

**Comment by juliayakovlev** [comment]
Claude failure:
```
Error: Action failed with error: SDK execution error: Error: Claude Code returned an error result: Credit balance is too low
```

File: sdcm/cluster_gce.py (hunk 4 above)
Comment (juliayakovlev) [reply sdcm/cluster_gce.py]: done

**Comment by juliayakovlev** [comment]
@fruch please, review

File: sdcm/cluster_gce.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (fruch) [reply sdcm/cluster_gce.py]: @juliayakovlev  didn't we talk about dropping the cleanup from here for now ?

File: sdcm/cluster_gce.py [hunk 7]
```diff
@@ -24,8 +24,9 @@
 
 from sdcm import cluster
 from sdcm.provision.gce.provisioner import GceProvisioner
+from sdcm.provision.gce.instance_provider import _is_zone_exhausted
```
Comment (dimakr) [inline sdcm/cluster_gce.py:29]: Maybe better rename this one (get rid of underscore) to no have import of a private name

File: defaults/test_default.yaml (hunk 5 above)
Comment (dimakr) [reply defaults/test_default.yaml:216]: It doesn't affect AWS logic, anyconfig.load merges aws_config.yaml value over test_default.yaml.

But shouldn't we keep this enabled for gce as well? (no other backend - Azure, OCI, Docker - uses this flag anyway) I.e. maybe we need this enabled globally to have resilient provisioning for gce by default? @juliayakovlev

File: test-cases/artifacts/ami.yaml [hunk 8]
```diff
@@ -13,3 +13,4 @@ scylla_linux_distro: 'centos'
 test_duration: 60
 user_prefix: 'artifacts-ami'
 aws_fallback_to_next_availability_zone: true
+fallback_to_next_availability_zone: true
```
Comment (dimakr) [inline test-cases/artifacts/ami.yaml:15]: Fallback param here and in other AWS-backend artifacts tests is redundant (as it is enabled by default in aws_config.yaml).
And if we decide to make the  `fallback_to_next_availability_zone` globally enabled, then it is redundant in all other artifacts test configs.

File: test-cases/artifacts/ami.yaml [hunk 9]
```diff
@@ -13,3 +13,4 @@ scylla_linux_distro: 'centos'
 test_duration: 60
 user_prefix: 'artifacts-ami'
 aws_fallback_to_next_availability_zone: true
```
Comment (dimakr) [inline test-cases/artifacts/ami.yaml]: Nothing to change here, just a note that this PR needs rebase as this state is stale - PR #14724 already removed this from aws test configs.

File: test-cases/artifacts/ami.yaml (hunk 8 above)
Comment (juliayakovlev) [reply test-cases/artifacts/ami.yaml:15]: The `fallback_to_next_availability_zone` parameter in the artifact YAML is currently only relevant for the GCE backend. Because this fallback mechanism is strictly supported in tests utilizing a 1-node cluster, it is set to False in test_defaults.yaml and explicitly enabled (True) only within the GCE artifact tests.

File: test-cases/artifacts/ami.yaml (hunk 9 above)
Comment (juliayakovlev) [reply test-cases/artifacts/ami.yaml]: rebased

File: defaults/test_default.yaml (hunk 5 above)
Comment (fruch) [reply defaults/test_default.yaml:216]: @dkropachev this PR current only handle artifact tests, and not all what needed for handling all other cases.

once it would handle everything, it would be enabled by default for GCE. (not in this PR)

Comment (dimakr) [reply defaults/test_default.yaml:216]: > @dkropachev this PR

this was probably comment for me)
Got it

File: unit_tests/unit/test_gce_zone_fallback.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright (c) 2026 ScyllaDB
+
+from unittest.mock import patch
```
Comment (dimakr) [inline unit_tests/unit/test_gce_zone_fallback.py]: As was marked already by the CodeRabbit - this new module duplicates the fixtures and tests from `unit_tests/unit/provisioner/test_gce_zone_resolver.py` (some are exact duplicates, others are nearly identical).
We can either remove this new module or take out the corresponding tests from the test_gce_zone_resolver.py

Comment (juliayakovlev) [reply unit_tests/unit/test_gce_zone_fallback.py]: removed

**Review by dimakr — APPROVED** [review]
LGTM

**Review by fruch — APPROVED** [review]
